test/opf/sdpwrm.jl

```diff
@@ -0,0 +1,222 @@
+        "tol_gap_rel"    => 1e-14,
+        "tol_feas"       => 1e-14,
+        "tol_infeas_rel" => 1e-14,
+        "tol_ktratio"    => 1e-14,
+    )
+    opf = OPFGenerator.build_opf(SDPWRMPowerModel, data, solver; T=T)
+    # set_silent(opf.model)
+    OPFGenerator.solve!(opf)
+    res = OPFGenerator.extract_result(opf)
+
+    _test_sdpwrm_DualFeasibility(data, res)
+
+    return nothing
+end
+
+function _get_sym(i, j, v, N)
+    # return an N * N sparse symmetric matrix where the (i, j) entry is v
+    return Symmetric(sparse([i], [j], v, N, N))
+end
+
+function _get_skew_sym(i, j, v, N)
+    # return an N * N sparse skew-symmetric matrix where the (i, j) entry is v
+    return sparse([i, j], [j, i], [v, -v], N, N)
+end
+
+function _test_sdpwrm_DualFeasibility(data, res; atol=1e-6)
+    ref = PM.build_ref(data)[:it][:pm][:nw][0]
+    N = length(ref[:bus])
+    E = length(ref[:branch])
+    bus_loads = [
+        [ref[:load][l] for l in ref[:bus_loads][i]]
+        for i in 1:N
+    ]
+    # Bus-level data
+    bus_shunts = [
+        [ref[:shunt][s] for s in ref[:bus_shunts][i]]
+        for i in 1:N
+    ]
+    gs = [sum(shunt["gs"] for shunt in bus_shunts[i]; init=0.0) for i in 1:N]
+    bs = [sum(shunt["bs"] for shunt in bus_shunts[i]; init=0.0) for i in 1:N]
+
+    # Extract branch-level data
+    g = [PM.calc_branch_y(ref[:branch][e])[1] for e in 1:E]
+    b = [PM.calc_branch_y(ref[:branch][e])[2] for e in 1:E]
+    tr = [PM.calc_branch_t(ref[:branch][e])[1] for e in 1:E]
+    ti = [PM.calc_branch_t(ref[:branch][e])[2] for e in 1:E]
+    ttm  = abs2.(tr) + abs2.(ti)
+    g_fr = [ref[:branch][e]["g_fr"] for e in 1:E]
+    g_to = [ref[:branch][e]["g_to"] for e in 1:E]
+    b_fr = [ref[:branch][e]["b_fr"] for e in 1:E]
+    b_to = [ref[:branch][e]["b_to"] for e in 1:E]
+    δθmin = [ref[:branch][e]["angmin"] for e in 1:E]
+    δθmax = [ref[:branch][e]["angmax"] for e in 1:E]
+
+    # Check dual feasibility for select buses and constraints
+    λp  = [res["solution"]["bus"]["$i"]["lam_kirchhoff_active"] for i in 1:N]
+    λq  = [res["solution"]["bus"]["$i"]["lam_kirchhoff_reactive"] for i in 1:N]
+    λpf = [res["solution"]["branch"]["$e"]["lam_ohm_active_fr"] for e in 1:E]
+    λqf = [res["solution"]["branch"]["$e"]["lam_ohm_reactive_fr"] for e in 1:E]
+    λpt = [res["solution"]["branch"]["$e"]["lam_ohm_active_to"] for e in 1:E]
+    λqt = [res["solution"]["branch"]["$e"]["lam_ohm_reactive_to"] for e in 1:E]
+
+    S = res["solution"]["S"]
+
+    μθ_lb = [res["solution"]["branch"]["$e"]["mu_va_diff_lb"] for e in 1:E]
+    μθ_ub = [-res["solution"]["branch"]["$e"]["mu_va_diff_ub"] for e in 1:E]
+
+    μ_w = [
+        res["solution"]["bus"]["$i"]["mu_w_lb"] + res["solution"]["bus"]["$i"]["mu_w_ub"]
+        for i in 1:N
+    ]
+
+    # Check dual constraint corresponding to `wr` variables
+    AR = sum(
+        (-gs[i] * λp[i] + bs[i] * λq[i]) * _get_sym(i, i, 1, N)
+        for i in 1:N;
+        init=0
+    )
+    + sum(
```

Comment:
This line break means that everything after `+ sum(...)` will not be included in the definition of `AR`

Reply:
Similar comment as above: while I don't know exactly how julia will end up computing the `sum` under the hood, you're summing a large number of very sparse matrices, each of which has a single non-zero coefficient... Not great for performance because summing sparse matrices requires re-allocating memory to merge the sparsity structure.

It believe it would be more efficient to build the COO representation, then call `sparse(I, J, V, N, N)` where `I, J, V` are 3 vectors containing the row index / column index / value of each non-zero entry. Note that, when converting from COO to CSC format, all duplicate entries are summed. See julia docs here: https://docs.julialang.org/en/v1/stdlib/SparseArrays/#Sparse-Vector-and-Matrix-Constructors